tonic/src/transport/server/mod.rs

```diff
@@ -80,6 +81,10 @@ pub trait NamedService {
     const NAME: &'static str;
 }
 
+impl<S: NamedService, T> NamedService for Either<S, T> {
+    const NAME: &'static str = S::NAME;
```

Comment:
This will only match if the `S` service name matches but doesn't work for T?

Reply:
Yeah.

A better alternative here would probably be:
`impl <S:NamedService> NamedService for Either<S, Empty Service>`.

Reply:
I think that having a `const` here limits us terribly because it needs to evaluate the service name at compile time while Either is a runtime enum trick. Because of that it doesn't really match very well.

I want to say warp has the same issue? There is one possible solution and that may be boxing things but I am not exactly sure how that would work with the const since you'd still need to define the `NAME` type at compile time even if you want to trait object.

Reply:
@LucioFranco Sorry for the inactivity and sorry I don't follow.

I've added an the following example which toggles a service based on a command line argument.

```rust
    let args: Vec<String> = env::args().collect();
    let enabled = args.get(1) == Some(&"enabled".to_string());

    let addr = "[::1]:50051".parse().unwrap();
    let greeter = MyGreeter::default();

    let optional_service = if enabled {
        println!("MyGreeter enabled");
        Some(GreeterServer::new(greeter))
    } else {
        println!("MyGreeter disabled");
        None
    };

    println!("GreeterServer listening on {}", addr);

    Server::builder()
        .add_optional_service(optional_service)
        .serve(addr)
        .await?;
```
and it seems to work fine.

Reply:
Try adding a service after, I think that will cause the issue I was talking about.

Reply:
So I just gave it another go:
```rust
#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    let enabled = args.get(1) == Some(&"enabled".to_string());

    // Set-up services
    let switch_service = if enabled {
        println!("Switch enabled");
        Some(SwitchableServer::new(MySwitch::default()))
    } else {
        println!("Switch disabled");
        None
    };
    let persistent_service = PersistentServer::new(MyPersistent::default());

    // Build server
    let addr = "[::1]:50051".parse().unwrap();
    Server::builder()
        .add_optional_service(switch_service)
        .add_service(persistent_service)
        .serve(addr)
        .await?;

    Ok(())
}
```
and still seems to work fine.

![Screenshot from 2020-07-01 21-22-15](https://user-images.githubusercontent.com/26074448/86287912-fd99b980-bbe0-11ea-9304-f3f740ed12f3.png)

Reply:
Right, what I don't think will work is if the Either enum is on the `T` generic, since that doesn't even have a bound? I can try to take a look tomorrow but in my head the `T` service here should not work.

Reply:
@LucioFranco okey. If you give it the green light I'll go ahead and add docs (tests) and the method on the `Router`.

Reply:
So another issue I see here (that may be totally minor and we may want to document) is that this service will catch the services with this name. So doing funky stuff around enabling two variations of the same service will cause issues.

Reply:
@LucioFranco Yeah I thought this might be the case, which is why I made two explicitly different services [here](https://github.com/hyperium/tonic/pull/275#discussion_r448596956).

I'll try paraphrase the phenomena in the documentation as best I can.